src/binaryen-c.h

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright 2016 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//================
+// Binaryen C API
+//
+// The first part of the API lets you create modules and their parts.
+// The second part of the API lets you perform operations on modules.
+//================
+
+#ifndef binaryen_h
+#define binaryen_h
+
+#include <stddef.h>
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+//
+// ========== Module Creation ==========
+//
+
+// BinaryenIndex
+//
+// Used for internal indexes and list sizes.
+
+typedef uint32_t BinaryenIndex;
+
+// Core types (call to get the value of each; you can cache them, they
+// never change)
+
+typedef uint32_t BinaryenType;
+
+BinaryenType BinaryenNone(void);
+BinaryenType BinaryenInt32(void);
+BinaryenType BinaryenInt64(void);
+BinaryenType BinaryenFloat32(void);
+BinaryenType BinaryenFloat64(void);
+
+// Modules
+//
+// Modules contain lists of functions, imports, exports, function types. The
+// Add* methods create them on a module. The module owns them and will free their
+// memory when the module is disposed of.
+//
+// Expressions are also allocated inside modules, and freed with the module. They
+// are not created by Add* methods, since they are not added directly on the
+// module, instead, they are arguments to other expressions (and then they are
+// the children of that AST node), or to a function (and then they are the body
+// of that function).
+//
+// A module can also contain a function table for indirect calls, a memory,
+// and a start method.
+
+typedef void* BinaryenModuleRef;
+
+BinaryenModuleRef BinaryenModuleCreate();
+void BinaryenModuleDispose(BinaryenModuleRef module);
+
+// Function types
+
+typedef void* BinaryenFunctionTypeRef;
+
+BinaryenFunctionTypeRef BinaryenAddFunctionType(BinaryenModuleRef module, const char* name, BinaryenType result, BinaryenType* paramTypes, BinaryenIndex numParams);
+
+// Literals. These are passed by value.
+
+struct BinaryenLiteral {
+  int type; // size of enum in c++
```

Comment:
Hmm, yeah, we should do that. Also it might save space since it fits in a byte.

However I think this might need to wait, as the gcc in Ubuntu 14.04 LTS has a bug on enum classes with int8_t that I've seen hit (e.g. in SpiderMonkey). Unless we do int8_t for the storage (no enum class, just plain int8_t) and an enum for the values, and do some casting, which maybe isn't so bad since it's only done internally in wasm.h?